instant-xml/src/lib.rs

```diff
@@ -181,8 +200,151 @@ pub struct FieldContext<'xml> {
+    }
+}
+
+pub struct Deserializer<'xml> {
+    parser: XmlParser<'xml>,
+    namespaces: HashMap<&'xml str, &'xml str>,
+    tag_attributes: Vec<(&'xml str, &'xml str)>,
+}
+
+impl<'xml> Deserializer<'xml> {
+    pub fn new(input: &'xml str) -> Self {
+        Self {
+            parser: XmlParser::new(input),
+            namespaces: std::collections::HashMap::new(),
+            tag_attributes: Vec::new(),
+        }
+    }
+
+    pub fn peek_next_tag(&mut self) -> Result<Option<XmlRecord>, Error> {
+        self.parser.peek_next_tag()
+    }
+
+    pub fn verify_namespace(&self, namespace_to_verify: &str) -> bool {
+        self.namespaces.get(namespace_to_verify).is_some()
+    }
+
+    pub fn peek_next_attribute(&self) -> Option<&(&'xml str, &'xml str)> {
+        self.tag_attributes.last()
+    }
+
+    pub fn deserialize_struct<V>(
+        &mut self,
+        visitor: V,
+        name: &str,
+        namespaces: &HashMap<&'xml str, &'xml str>,
+    ) -> Result<V::Value, Error>
+    where
+        V: Visitor<'xml>,
+    {
+        let new_namespaces = namespaces
+            .iter()
+            .filter(|(k, v)| self.namespaces.insert(k, v).is_none())
+            .collect::<Vec<_>>();
+
+        self.process_open_tag(name, namespaces)?;
+        let ret = visitor.visit_struct(self)?;
+
+        self.check_close_tag(name)?;
+        let _ = new_namespaces
+            .iter()
+            .map(|(k, _)| self.namespaces.remove(*k));
+
+        Ok(ret)
+    }
+
+    fn deserialize_bool<V>(&mut self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: Visitor<'xml>,
+    {
+        self.parser.next();
+        if let Some(Ok(XmlRecord::Element(v))) = self.parser.next() {
+            let ret = visitor.visit_str(v);
+            self.parser.next();
+            ret
+        } else {
+            Err(Error::UnexpectedValue)
+        }
+    }
+
+    fn deserialize_attribute<V>(&mut self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: Visitor<'xml>,
+    {
+        match self.tag_attributes.pop() {
+            Some((_, value)) => visitor.visit_str(value),
+            None => Err(Error::UnexpectedEndOfStream),
+        }
+    }
+
+    fn process_open_tag(
+        &mut self,
+        name: &str,
+        namespaces: &HashMap<&'xml str, &'xml str>,
+    ) -> Result<(), Error> {
+        if let Some(Ok(XmlRecord::Open(item))) = self.parser.next() {
+            if item.key == name {
+                for (k, v) in item.namespaces.unwrap() {
+                    if let Some(item) = namespaces.get(k) {
+                        if *item != v {
+                            return Err(Error::UnexpectedPrefix);
+                        }
+                    } else {
+                        return Err(Error::MissingdPrefix);
+                    }
+                }
+
+                self.tag_attributes = item.attributes;
+                return Ok(());
+            }
+        }
+
+        Err(Error::UnexpectedValue)
```

Comment:
Untested, but this should be more like this:

```suggestion
        let item = match self.parser.next() {
            Some(Ok(XmlRecord::Open(item))) if item.key == name => item,
            _ => return Err(Error::UnexpectedValue),
        };

        for (k, v) in item.namespaces.unwrap() {
            let item = match item.namespaces.get(k) {
                Some(item) if *item == v => {
                     self.tag_attributes = item.attributes;
                     return Ok(());
                }
                Some(item) => return Err(Error::UnexpectedPrefix),
                None => return Err(MissingPrefix),
            }
```